pkg/apply/filter/inventory-policy-apply-filter.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2021 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package filter
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+	"sigs.k8s.io/cli-utils/pkg/object"
+)
+
+// InventoryPolicyApplyFilter implements ValidationFilter interface to determine
+// if an object should be applied based on the cluster object's inventory id,
+// the id for the inventory object, and the inventory policy.
+type InventoryPolicyApplyFilter struct {
+	Client    dynamic.Interface
+	Mapper    meta.RESTMapper
+	Inv       inventory.InventoryInfo
+	InvPolicy inventory.InventoryPolicy
+}
+
+// Name returns a filter identifier for logging.
+func (ipaf InventoryPolicyApplyFilter) Name() string {
+	return "InventoryPolicyApplyFilter"
+}
+
+// Filter returns true if the passed object should be filtered (NOT applied) and
+// a filter reason string; false otherwise. Returns an error if one occurred
+// during the filter calculation
+func (ipaf InventoryPolicyApplyFilter) Filter(obj *unstructured.Unstructured) (bool, string, error) {
+	if obj == nil {
+		return true, "missing object", nil
+	}
+	// Object must be retrieved from the cluster to get the inventory id.
+	clusterObj, err := ipaf.getObject(object.UnstructuredToObjMetaOrDie(obj))
```

Comment:
Does this add an additional `Get` for every resource, or does it just move the existing `Get` from inside the apply task to a filter? 
Regardless, I'm wondering if we should look at ways where the object from the cluster are provided to the filter, instead of it being fetched by the filter itself.  This approach doesn't seem to scale well with additional filters.

Reply:
It moves the existing `Get` to the filter. This pattern is necessary since we want to pass in the local file to filters. For example, if we implement the `local-config` filtering in a `ValidationFilter` we have to pass the local file--not the file fetched from the cluster. It is not even guaranteed that there is a cluster object (it won't exist for `local-config` filtered files). The only way this works is using local files.